### PR TITLE
CSP-684 - Add GET Stage Apprentice endpoint 

### PR DESCRIPTION
**Pull request by ravihejmadi**
Added Controller, Handler, Validator, Result, Repo & Tests

### CODE REVIEW

File: src/SFA.DAS.AANHub.Domain/Interfaces/IAanDataContext.cs [hunk 1]
```diff
@@ -12,6 +12,7 @@ public interface IAanDataContext
         DbSet<Employer> Employers { get; }
         DbSet<Partner> Partners { get; }
         DbSet<Admin> Admins { get; }
+        DbSet<StagedApprentice> StagedApprentice { get; }
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Domain/Interfaces/IAanDataContext.cs:15]: Rename this to `StagedApprentices`, to be plural

File: src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        public long Uln { get; set; }
+        public long ApprenticeshipId { get; set; }
+        public string EmployerName { get; set; } = string.Empty;
+        public DateTime? StartDate { get; set; }
+        public DateTime? EndDate { get; set; }
+        public long TrainingProviderId { get; set; }
+        public string TrainingProviderName { get; set; } = string.Empty;
+        public string TrainingCode { get; set; } = null!;
+        public string TrainingCourseOption { get; set; } = string.Empty;
+        public string StandardUId { get; set; } = string.Empty;
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs:19]: These are nullable right, so they should be declared nullable and empty string assignment should be removed.

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿using SFA.DAS.AANHub.Domain.Entities;
+
+namespace SFA.DAS.AANHub.Application.StagedApprentices.Queries
+{
+    public class GetStagedApprenticeResult
+    {
+        public long Uln { get; set; }
+        public long ApprenticeshipId { get; set; }
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs:8]: Missing fields like first name, last name etc?

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeQueryValidator.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class GetStagedApprenticeQueryValidator : AbstractValidator<GetStagedApprenticeQuery>
+    {
+        public GetStagedApprenticeQueryValidator()
+        {
+            RuleFor(a => a.LastName)
+                .NotEmpty();
+
+            RuleFor(a => a.DateOfBirth)
+            .NotEmpty();
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeQueryValidator.cs:13]: Alignment

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeQuery.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public GetStagedApprenticeQuery(string lastname, DateTime dateofbirth, string email)
+        {
+            LastName = lastname;
+            DateOfBirth = dateofbirth;
+            Email = email;
+        }
+        public string LastName { get; }
+        public DateTime DateOfBirth { get; }
+        public string Email { get; }
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeQuery.cs:16]: Move properties above constructor

File: src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                TrainingCode = "TestTrainingCode",
+                StandardUId = "TestStandardUId"
+            };
+
+            stagedApprenticesReadRepositoryMock.Setup(a => a.GetStagedApprentice(lastname, dateofbirth, email)).ReturnsAsync(stagedApprentice);
+            var sut = new GetStagedApprenticeMemberQueryHandler(stagedApprenticesReadRepositoryMock.Object);
+
+            var result = await sut.Handle(new GetStagedApprenticeQuery(lastname, dateofbirth, email), new CancellationToken());
+
+            Assert.AreEqual(stagedApprentice.Uln, result.Result.Uln);
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs:38]: Could have used fluent assertion extension IsEquivalentTo

File: src/SFA.DAS.AANHub.Api/Controllers/StagedApprenticesController.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        [HttpGet]
+        [Route("/StagedApprentice/lastname={lastname}&dateofbirth={dateofbirth}&email={email}")]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(string), StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(typeof(GetStagedApprenticeResult), StatusCodes.Status200OK)]
+        public async Task<IActionResult> GetStagedApprentice(string lastname, DateTime dateofbirth, string email)
+        {
+            _logger.LogInformation("AAN Hub API: Received command to get StagedApprentice by LastName: {lastname}, DateOfBirth: {dateofbirth} and Email: {email}", lastname, dateofbirth, email);
+
+            var response = await _mediator.Send(new GetStagedApprenticeQuery(lastname, dateofbirth, email));
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api/Controllers/StagedApprenticesController.cs:37]: If the apprentice is not found the handler will return null which is not handled here. We should add NotFound to the ProducesResponseType as well.

File: src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs (hunk 6 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs:38]: Applied Fluent Assertions

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeQuery.cs (hunk 5 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeQuery.cs:16]: Moved further up. I was think about this when writing but then choose not

File: src/SFA.DAS.AANHub.Api/Controllers/StagedApprenticesController.cs (hunk 7 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Api/Controllers/StagedApprenticesController.cs:37]: Done

File: src/SFA.DAS.AANHub.Domain/Interfaces/IAanDataContext.cs (hunk 1 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Domain/Interfaces/IAanDataContext.cs:15]: image
Changed to plural but am getting the above error. This is after I merged **CSP-683** into `main `and rebased from `main `(I can see the StagedApprentice.sql file in the `.sln` now).

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs (hunk 3 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs:8]: Closing this based on @alandburns response on comms

File: src/SFA.DAS.AANHub.Domain/Interfaces/IAanDataContext.cs (hunk 1 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Domain/Interfaces/IAanDataContext.cs]: Thanks for your help on this @DevenKShah. Closing this now

File: src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+using SFA.DAS.AANHub.Domain.Entities;
+using SFA.DAS.AANHub.Domain.Interfaces.Repositories;
+
+namespace SFA.DAS.AANHub.Application.UnitTests.StagedApprentices.Queries
+{
+    [TestFixture]
+    public class GetStagedApprenticeQueryHandlerTests
+    {
+        [Test]
+        public async Task Handle_GetStagedApprenticeMember()
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs:14]: Add a test to return a ValidatedResponse with null result.

File: src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeResultTest.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        [Test, AutoMoqData]
+        public void StagedApprentice_GetStagedApprentice_StagedApprenticeIsNull()
+        {
+            StagedApprentice? stagedApprentice = null;
+            var response = (GetStagedApprenticeResult?)stagedApprentice!;
+            Assert.That(response, Is.Null);
+        }
+
+        //[Test]
+        //[AutoData]
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeResultTest.cs:43]: Delete commented code

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
+﻿using SFA.DAS.AANHub.Domain.Entities;
+
+namespace SFA.DAS.AANHub.Application.StagedApprentices.Queries
+{
+    public class GetStagedApprenticeResult
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs:5]: GetStagedApprenticeQueryResult

File: src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    public class StagedApprentice
+    {
+        public int Id { get; set; }
+        public string FirstName { get; set; } = null!;
+        public string LastName { get; set; } = null!;
+        public string Email { get; set; } = null!;
+        public DateTime? DateOfBirth { get; set; }
+        public long Uln { get; set; }
+        public long ApprenticeshipId { get; set; }
+        public string?EmployerName { get; set; } = string.Empty;
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs:12]: Add space after ?

Comment (alandburns) [reply src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs:12]: Uln and Apprenticeshipid are optional, and will only be set if known.

File: src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeResultTest.cs (hunk 9 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeResultTest.cs]: Deleted

File: src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs (hunk 10 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application/StagedApprentices/Queries/GetStagedApprenticeResult.cs]: renamed in both Application and Tests proj

File: src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs (hunk 8 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application.UnitTests/StagedApprentices/Queries/GetStagedApprenticeQueryHandlerTests.cs:14]: Added Test

File: src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs (hunk 11 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs]: Amended this issue and a few others. Re Alan's comment  re Uln and Apprenticeshipid - Do I make the `long` nullable? Haven't seen a `nullable` `int` or `long` ever!

Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Domain/Entities/StagedApprentice.cs]: @ravihejmadi yes set those fields to nullable please

**Review by DevenKShah — DISMISSED** [review]

**Review by YuvrajPatil-DFE — APPROVED** [review]
